### DISCUSSION TITLE
Low poly primitive pyramid 🔺 for agent visualisation 🔥🔥

### DISCUSSION THREAD

**Discussion by zeyus** [FLAMEGPU/FLAMEGPU2#881] [Show and tell]
- [Intro](#intro)
- [Specs](#specs)
- [Using the model](#using-the-model)
- [Screenshots](#screenshots)
- [License](#license)
- [Download](#download)

## Intro<a id="intro"></a>

I only have a GTX1070, and my simulation is going to have a lot of agents, obviously the visualization part isn't necessary, but it's very nice to have, also for figures in the paper.

Anyway, as such, I made the most simple 3D agent model I could imagine in blender, and anyone is welcome to use it if they would like.

## Specs<a id="specs"></a>

| Property | Value |
|-----------|-------:|
| Vertices | 4 |
| Edges | 6 |
| Faces/polygons | 4 |
| Face corners | 12 |

## Using the model<a id="using-the-model"></a>

Extract the `primitive_pyramid.obj` file to somewhere you can reference from your code, and then as per the [flamegpu agent model documentation](https://docs.flamegpu.com/guide/visualisation/visualising-agents.html#agent-model), just call the `AgentVis::setModel` function of your initialised `AgentVis` class.

Example:

```python
import pyflamegpu

model: pyflamegpu.ModelDescription = pyflamegpu.ModelDescription('test_model')
agent: pyflamegpu.AgentDescription = model.newAgent('test_agent')
env: pyflamegpu.EnvironmentDescription = model.Environment()
simulation: pyflamegpu.CUDASimulation = pyflamegpu.CUDASimulation(model)
visualisation: pyflamegpu.ModelVis = simulation.getVisualisation()
vis_agent: pyflamegpu.AgentVis = visualisation.addAgent('test_agent')

# specify the custom low poly model
vis_agent.setModel('./path/to/primitive_pyramid.obj')
# populate simulation etc
# ...
simulation.simulate()
```

## Screenshots<a id="screenshots"></a>

Model in blender:
<img width="403" alt="image" src="https://user-images.githubusercontent.com/75656/177001265-2e7e9170-4ace-4a39-8703-08c6153132cb.png">

Screenshot from a flamegpu simulation:
![screenshot](https://user-images.githubusercontent.com/75656/177001417-78802f7a-e3cc-4bde-8c08-0fee719a7391.png)

Updated version with actual equilateral triangular pyramid + arrowhead (for directional indication):

![triangular_pyramids](https://user-images.githubusercontent.com/75656/177033583-5c48858b-27f1-4f77-bdc2-b22809f98841.png)

## Licence<a id="license"></a>

> The MIT License (MIT)
> 
> Copyright (c) 2022 zeyus (https://github.com/zeyus)
> 
> Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
> 
> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
> 
> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

## Download<a id="download"></a>

Download updated version with an arrowhead model and a (proper) equilateral triangular pyramid (including blender file and exported `.obj`): 
[pyramids_equi_and_arrow.zip](https://github.com/FLAMEGPU/FLAMEGPU2/files/9034433/pyramids_equi_and_arrow.zip)

Download original eyeballed triangular pyramid model (including blender file and exported `.obj`): [Primitive_Pyramid.zip](https://github.com/FLAMEGPU/FLAMEGPU2/files/9033025/Primitive_Pyramid.zip)

**Comment by Robadob** [comment]
Great idea creating a 4 triangle model, not something I had considered. If you want it merged put a PR into [the visualiser's repo](https://github.com/FLAMEGPU/FLAMEGPU2-visualiser)

* A copy of the `.obj` would go in `resources/`
* Add the path of the new `.obj` file to `src/CMakeLists.txt:316`
* Add it's resource path to `include/flamegpu/visualiser/config/Stock.h` (you should credit yourself here in the comment, and possibly worth stating it's orientation).

I think that's all which is required to properly integrate it, but you may also want to update the [userguide](https://github.com/FLAMEGPU/FLAMEGPU2-docs):

* Add it to the list at `src/guide/visualisation/visualising-agents.rst:39`

There could be an argument for making a none-equilateral version, so it more clearly shows it's orientation (like an arrow-head).

**Reply by zeyus** [reply]
yeah I actually thought about the non-equilateral version myself but for my sim direction doesn't matter. I'll make a PR and I can create a second version that has a "front" (also, to be honest, it's not perfectly equilateral, I just eyeballed it, but I'll make a proper equilateral version). I just made it so Z+ was up, but I'll make sure to note it.

Expect an incoming PR within the next few hours (obviously no pressure to merge anytime soon).
